R/leave_location_out_cv.R

```diff
@@ -0,0 +1,193 @@
+  ## derive them later when they are needed.
+
+  split_objs$splits <- map(split_objs$splits, rm_out)
+
+  ## Save some overall information
+
+  cv_att <- list(v = v)
+
+  new_rset(
+    splits = split_objs$splits,
+    ids = split_objs[, grepl("^id", names(split_objs))],
+    attrib = cv_att,
+    subclass = c("leave_location_out_cv", "rset")
+  )
+}
+
+llo_splits <- function(data, location, v, pool = 0.1) {
+
+  n <- nrow(data)
+  locations <- tibble::tibble(
+    idx = seq_len(n),
+    location = location
+  )
+
+  folds <- tibble::tibble(
+    location = unique(locations$location),
+    fold = sample(rep(seq_len(v), length.out = length(unique(locations$location))))
+  )
+
+  locations <- merge(locations, folds, by = c("location"))
+  indices <- split_unnamed(locations$idx, locations$fold)
+  indices <- lapply(indices, default_complement, n = n)
+  split_objs <- purrr::map(
+    indices,
+    make_splits,
+    data = data,
+    class = "llo_split"
+  )
+  tibble::tibble(
+    splits = split_objs,
+    id = names0(length(split_objs), "Fold")
+  )
+}
+
+#' @export
+print.leave_location_out_cv <- function(x, ...) {
+    cat("# ", pretty(x), "\n")
+    class(x) <- class(x)[!(class(x) %in% c("leave_location_out_cv", "rset"))]
+    print(x, ...)
+}
+
+#' Create or Modify Location Variables
+#'
+#' @details
+#' For categorical inputs, the function will find levels of `x` than
+#'   occur in the data with percentage less than `pool`. The values from
+#'   these groups will be pooled into a single category (if they make up
+#'   a greater than `pool` percentage of the data) or combined with the smallest
+#'   extant category (if not).
+#'
+#' @param x An input vector.
+#' @param pool A proportion of data used to determine if a particular group is
+#'   too small and should be pooled into another group.
+#' @return  A factor vector.
+make_location <- function(x, pool = .1) {
+
+  n <- length(x)
+  x <- factor(x)
+  xtab <- sort(table(x))
+  pcts <- xtab / n
+
+  if (all(pcts < pool) || sum(pcts > pool) == 1) {
+    rlang::abort(c(
+      "Fewer than two locations had enough data to use without pooling.",
+      "Consider providing a less granular location variable."
+    ))
+  }
+
+  ## Small groups will all be thrown into either a single pooled location,
+  ## or appended to the smallest extant grouping
+  if (any(pcts < pool)) {
+    x[x %in% names(pcts)[pcts < pool]] <- NA
+  }
+
+  default_name <- ".pooled_locations"
+  replacement_name <- default_name
+
+  ## If the combined location itself wouldn't be bigger than `pool`,
+  ## combine them with the smallest group that does:
+  if ((sum(pcts < pool) / n) < pool) {
+      replacement_name <- xtab[which(pcts > pool)]
+      replacement_name <- names(replacement_name[which.min(replacement_name)])
+    rlang::warn(c(
+      paste0("Combining small locations into a new group would create a ",
+             "group smaller than ", round(100 * pool), "% of the data."),
```

Comment:
It makes sense to me that, if you have a single group that's smaller than `pool`, you don't want it to just be re-labeled and left as a small group. But at the same time, the "add it to the smallest group" behavior might not be desirable.